hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlExecuteMessageTask.java

```diff
@@ -0,0 +1,124 @@
+import com.hazelcast.client.impl.protocol.task.AbstractCallableMessageTask;
+import com.hazelcast.instance.impl.Node;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.sql.SqlQuery;
+import com.hazelcast.sql.impl.SqlInternalService;
+import com.hazelcast.sql.impl.SqlResultImpl;
+import com.hazelcast.sql.impl.SqlServiceImpl;
+
+import java.security.Permission;
+
+/**
+ * SQL query execute task.
+ */
+public class SqlExecuteMessageTask extends AbstractCallableMessageTask<SqlExecuteCodec.RequestParameters> {
+    public SqlExecuteMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected Object call() throws Exception {
+        try {
+            SqlQuery query = new SqlQuery(parameters.sql);
+
+            if (parameters.parameters != null && !parameters.parameters.isEmpty()) {
+                for (Data param : parameters.parameters) {
+                    query.addParameter(serializationService.toObject(param));
+                }
+            }
+
+            query.setTimeoutMillis(parameters.timeoutMillis);
+            query.setCursorBufferSize(parameters.cursorBufferSize);
+
+            SqlServiceImpl sqlService = nodeEngine.getSqlService();
+
+            SqlResultImpl cursor = (SqlResultImpl) sqlService.query(query);
+
+            SqlPage page = sqlService.getInternalService().getClientStateRegistry().registerAndFetch(
+                endpoint.getUuid(),
+                cursor,
+                parameters.cursorBufferSize,
+                serializationService
+            );
+
+            return new SqlExecuteResponse(
+                cursor.getQueryId(),
+                cursor.getRowMetadata(),
+                page,
+                null
+            );
+        } catch (Exception e) {
+            SqlError error = SqlClientUtils.exceptionToClientError(e, nodeEngine.getLocalMember().getUuid());
+
+            return new SqlExecuteResponse(null, null, null, error);
+        }
+    }
+
+    @Override
+    protected SqlExecuteCodec.RequestParameters decodeClientMessage(ClientMessage clientMessage) {
+        return SqlExecuteCodec.decodeRequest(clientMessage);
+    }
+
+    @Override
+    protected ClientMessage encodeResponse(Object response) {
+        SqlExecuteResponse response0 = (SqlExecuteResponse) response;
+
+        return SqlExecuteCodec.encodeResponse(
+            response0.getQueryId(),
+            response0.getRowMetadata(),
+            response0.getPage(),
+            response0.getError()
+        );
+    }
+
+    @Override
+    public String getServiceName() {
+        return SqlInternalService.SERVICE_NAME;
+    }
+
+    @Override
+    public String getDistributedObjectName() {
+        return null;
+    }
+
+    @Override
+    public String getMethodName() {
+        return "execute";
```

Comment:
for the method name, we usually use the user API. And note that related tests are on the enterprise repo. Make sure that you have `SecurityInterceptor` and permission tests for the new use case.

Reply:
I can change it to `query`, but what is the goal of this? From the user perspective, he executes a query, but internally it is translated into several different commands, that are not directly mapped to the public API. Moreover, the result of this operation contains internal objects, which also highlights that this is not a direct API call.

Reply:
Well, we don't really have a clear answer. In the end, whatever you choose must be documented clearly. We select it as same as the public method name to make it more intuitive. Not sure what to do when it is multiple requests per API call. 
Just be aware that what you choose here is also a public API and needs to be documented.